SD_Card_Driver/Core/Src/main.cpp

```diff
@@ -0,0 +1,389 @@
+  * @file           : main.c
+  * @brief          : Main program body
+  ******************************************************************************
+  * @attention
+  *
+  * Copyright (c) 2022 STMicroelectronics.
+  * All rights reserved.
+  *
+  * This software is licensed under terms that can be found in the LICENSE file
+  * in the root directory of this software component.
+  * If no LICENSE file comes with this software, it is provided AS-IS.
+  *
+  ******************************************************************************
+  */
+/* USER CODE END Header */
+/* Includes ------------------------------------------------------------------*/
+#include "main.h"
+#include "LOS_D_sd.hpp"
+
+/* Private includes ----------------------------------------------------------*/
+/* USER CODE BEGIN Includes */
+#include <stdio.h>
+#include <string.h>
+#include <stdarg.h> //for va_list var arg functions
+/* USER CODE END Includes */
+
+/* Private typedef -----------------------------------------------------------*/
+/* USER CODE BEGIN PTD */
+
+/* USER CODE END PTD */
+
+/* Private define ------------------------------------------------------------*/
+/* USER CODE BEGIN PD */
+#define PRINT_BUF_SIZE 128
+/* USER CODE END PD */
+
+/* Private macro -------------------------------------------------------------*/
+/* USER CODE BEGIN PM */
+
+/* USER CODE END PM */
+
+/* Private variables ---------------------------------------------------------*/
+SPI_HandleTypeDef hspi2;
+
+UART_HandleTypeDef huart2;
+
+/* USER CODE BEGIN PV */
+
+/* USER CODE END PV */
+
+/* Private function prototypes -----------------------------------------------*/
+void SystemClock_Config(void);
+static void MX_GPIO_Init(void);
+static void MX_USART2_UART_Init(void);
+static void MX_SPI2_Init(void);
+/* USER CODE BEGIN PFP */
+void myprintf(const char *fmt, ...);
+/* USER CODE END PFP */
+
+/* Private user code ---------------------------------------------------------*/
+/* USER CODE BEGIN 0 */
+void myprintf(const char *fmt, ...)
+{
+  static char buffer[256];
+  va_list args;
+  va_start(args, fmt);
+  vsnprintf(buffer, sizeof(buffer), fmt, args);
+  va_end(args);
+
+  int len = strlen(buffer);
+  HAL_UART_Transmit(&huart2, (uint8_t*)buffer, len, -1);
+}
+
+void test()
+{
+	int res;
+	StorageDevice* storageDevice = &SDCard::getInstance();
+
+	/* Init storage device */
+	res = storageDevice->init();
+	if (res == 0) {
+		myprintf("* Initialized storage device\r\n");
+	} else {
+		myprintf("* Failed to initialize storage device! Error code = %d\r\n", res);
+		return;
+	}
+
+	const char* fileName = "test.txt";
+	const char* writeData = "test\ntest";
+	size_t writeDataSize = strlen(writeData);
+	uint8_t* readData;
+	size_t fileSize;
+
+	/* Check if file exists */
+	bool fileExist = storageDevice->checkExist(fileName);
+	if (fileExist) {
+		myprintf("* Test file \"%s\" exists. Please delete it and run the test again.\r\n", fileName);
+		goto end;

```

Comment:
The use of goto is generally a big no go as it confuses the logic flow when someone tries to read it. Probably goes against safety standards as well (but i'm not sure about that)

Reply:
Generally, goto's are a big nono. The barr_c standard gives a pretty good explanation.
https://barrgroup.com/sites/default/files/barr_c_coding_standard_2018.pdf